tree/ntuple/v7/src/RNTupleSerialize.cxx

```diff
@@ -0,0 +1,1363 @@
+      return R__FAIL("invalid envelope, too short");
+
+   auto bytes = reinterpret_cast<const unsigned char *>(buffer);
+   auto base = bytes;
+
+   std::uint16_t protocolVersionAtWrite;
+   std::uint16_t protocolVersionMinRequired;
+   bytes += DeserializeUInt16(bytes, protocolVersionAtWrite);
+   // RNTuple compatible back to version 1 (but not to version 0)
+   if (protocolVersionAtWrite < 1)
+      return R__FAIL("The RNTuple format is too old (version 0)");
+
+   bytes += DeserializeUInt16(bytes, protocolVersionMinRequired);
+   if (protocolVersionMinRequired > kEnvelopeCurrentVersion) {
+      return R__FAIL(std::string("The RNTuple format is too new (version ") +
+                                 std::to_string(protocolVersionMinRequired) + ")");
+   }
+
+   // We defer the CRC32 check to the end to faciliate testing of forward/backward incompatibilities
+   auto result = VerifyCRC32(base, bufSize - 4, crc32);
+   if (!result)
+      return R__FORWARD_ERROR(result);
+
+   return sizeof(protocolVersionAtWrite) + sizeof(protocolVersionMinRequired);
+}
+
+
+RResult<std::uint32_t> ROOT::Experimental::Internal::RNTupleSerializer::DeserializeEnvelope(
+   const void *buffer, std::uint32_t bufSize)
+{
+   std::uint32_t crc32;
+   return R__FORWARD_RESULT(DeserializeEnvelope(buffer, bufSize, crc32));
+}
+
+
+std::uint32_t ROOT::Experimental::Internal::RNTupleSerializer::SerializeRecordFramePreamble(void *buffer)
+{
+   // Marker: multiply the final size with 1
+   return SerializeInt32(1, buffer);
+}
+
+
+std::uint32_t ROOT::Experimental::Internal::RNTupleSerializer::SerializeListFramePreamble(
+   std::uint32_t nitems, void *buffer)
+{
+   if (nitems >= (1 << 28))
+      throw RException(R__FAIL("list frame too large: " + std::to_string(nitems)));
+
+   auto base = reinterpret_cast<unsigned char *>((buffer != nullptr) ? buffer : 0);
+   auto pos = base;
+   void** where = (buffer == nullptr) ? &buffer : reinterpret_cast<void**>(&pos);
+
+   // Marker: multiply the final size with -1
+   pos += RNTupleSerializer::SerializeInt32(-1, *where);
+   pos += SerializeUInt32(nitems, *where);
+   return pos - base;
+}
+
+std::uint32_t ROOT::Experimental::Internal::RNTupleSerializer::SerializeFramePostscript(
+   void *frame, std::int32_t size)
+{
+   if (size < 0)
+      throw RException(R__FAIL("frame too large: " + std::to_string(size)));
+   if (size < static_cast<std::int32_t>(sizeof(std::int32_t)))
+      throw RException(R__FAIL("frame too short: " + std::to_string(size)));
+   if (frame) {
+      std::int32_t marker;
+      DeserializeInt32(frame, marker);
+      if ((marker < 0) && (size < static_cast<std::int32_t>(2 * sizeof(std::int32_t))))
+         throw RException(R__FAIL("frame too short: " + std::to_string(size)));
+
+      SerializeInt32(marker * size, frame);
+   }
+   return 0;
+}
+
+ROOT::Experimental::RResult<std::uint32_t> ROOT::Experimental::Internal::RNTupleSerializer::DeserializeFrame(
```

Comment:
```suggestion
ROOT::Experimental::RResult<std::uint32_t> ROOT::Experimental::Internal::RNTupleSerializer::DeserializeFrameHeader(
```
as this deal "only" with the size and the number of element and not the payload.